File: src/xray.py
#!/usr/bin/env python3

"""
2025 Meme Supplier
memesupplierbusiness@gmail.com
Maintained by Meme Supplier
"""

import subprocess
import readline
import sys

version = "v1.4"

def choose():
    global file
    choice = input("Choice: ")

    match choice:
        case 'exit':
            sys.exit(0)
        case '0':
            sys.exit(0)
        case '1':
            file = "main.rs"
        case '2':
            file = "cmds.rs"
        case '3':
            file = "funcs.rs"
        case '4':
            file = "diskfmt.py"
        case '5':
            file = "ver.py"
        case '6':
            file = "xray.py"
        case '7':
            file = "install.sh"
        case '8':
            file = "launch.sh"
        case '9':
            file = "upgrade.sh"
        case '10':
            file = "uninstall.sh"
        case _:
            print("Invalid option! Choose again!")
            choose()

print(f"\nRusterminal Viewer {version}")
print("\nChoose file to view (select number):\nType \"0\" or \"exit\" to exit.\n")

files = ["1] main.rs",
         "2] cmds.rs",
         "3] funcs.rs",
         "4] diskfmt.py",
         "5] ver.py",
         "6] xray.py",
         "7] install.sh",
         "8] launch.sh",
         "9] upgrade.sh",
         "10] uninstall.sh\n"]

for line in files:
    print(line)

choose()
subprocess.run([f"nano ~/rusterminal/src/{file}"],shell=True)

File: src/xray.rs
use std::io::{self, Write};

use crate::funcs::run_shell_command;
use crate::logger::log;

static VERSION: &str = "2.6";

fn list() {
    let lines: [&str; 11] = [
        "1] src/main.rs",
        "2] src/cmds.rs",
        "3] src/funcs.rs",
        "4] src/logger.rs",
        "5] src/diskfmt.py",
        "6] src/sys/s_info.rs",
        "7] src/xray.rs",
        "8] installer/upgrade.sh",
        "9] launch.sh",
        "10] installer/uninstall.sh",
        "11] src/sys/s_vars.rs\n",
    ];

    for line in lines.iter() {
        println!("{line}")
    }
}

fn choose(input: &str) {
    let input = input.trim();

    let file = match input {
        "1" => Some("main.rs"),
        "2" => Some("cmds.rs"),
        "3" => Some("funcs.rs"),
        "4" => Some("logger.rs"),
        "5" => Some("diskfmt.py"),
        "6" => Some("s_info.rs"),
        "7" => Some("xray.rs"),
        "8" => Some("upgrade.sh"),
        "9" => Some("launch.sh"),
        "10" => Some("uninstall.sh"),
        "11" => Some("s_vars.rs"),
        _ => {
            println!("{input}: Invalid filename!");
            None
        }
    };

    if let Some(file) = file {
        log(&format!("xray::choose(): Viewing file: {file}"));

        if file == "upgrade.sh" || file == "uninstall.sh" {
            run_shell_command(&format!("nano ~/rusterminal/installer/{file}"))
        } else if file == "s_vars.rs" || file == "s_info.rs" {
            run_shell_command(&format!("nano ~/rusterminal/src/sys/{file}"))
        } else {
            run_shell_command(&format!("nano ~/rusterminal/src/{file}"))
        }
    }
}

pub fn main() {
    log(&format!(
        "xray::main(): Rusterminal Viewer version: {VERSION}"
    ));

    println!("\nRusterminal Viewer v{VERSION}\n\nChoose file to view (select number):\nType \"0\" or \"exit\" to exit.\n");
    list();

    print!("Choice: ");
    let mut input = String::new();

    io::stdout().flush().expect("Failed to flush");

    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read line");

    choose(&input);
    println!();
    run_shell_command("clear")
}
